Major update: dynamic anchors

thefliik opened the pull request:
Hi, and thanks for publishing this library!

I'm not sure if this PR will be desired, but I needed the ability to dynamically change the anchor associated with a popover so I built that out. This PR upstreams my changes. This is a breaking change, but one that simplifies the codebase (and, I would argue, the API) without any feature loss and with a feature added.

After this update, `SatPopover` has an `anchor` input property which can be provided one of `SatPopoverAnchor` or `ElementRef` or `HTMLElement`. Alternatively (similar to before), a `SatPopoverAnchor` directive exists which has a `[satPopoverAnchorFor]="contactPopover"` input setter which can be used to associate an anchor with a popover.

Example of new usage with `SatPopover#anchor`:

```typescript
<button satPopoverAnchor #anchor=satPopoverAnchor (click)="popover.toggle()">
  See Contact Details
</button>

<sat-popover #popover [anchor]="anchor" hasBackdrop>
  <app-contact-overview [contact]="myContact"></app-contact-overview>
</sat-popover>
```

_Note: Template reference variables, if placed on a generic html element, select the `HTMLElement`. When placed on an angular component, they select the angular component and there is no way to have them select the `HTMLElement`. Because of this, you must use the `satPopoverAnchor` directive if using an angular component as an anchor._

Because the above example uses a simple html `<button>` element, the above example _could_ be rewritten to remove `satPopoverAnchor`. For example, the following works (but if a `MatButton` were used, i.e. `<button mat-button>` from `@angular/material`, the following example would NOT work) :

```typescript
<button #anchor (click)="popover.toggle()">
  See Contact Details
</button>

<sat-popover #popover [anchor]="anchor" hasBackdrop>
  <app-contact-overview [contact]="myContact"></app-contact-overview>
</sat-popover>
```

Because of this potentially confusing detail, I think it's wise to continue using `satPopoverAnchor` in the documentation, even for examples where it isn't actually necessary.

Along with this PR, `SatPopoverAnchor` has been gutted and is now a simple directive that provides a handle to the associated `SatPopover`. `SatPopoverAnchor` is still required when using `SatPopoverHoverDirective`.

With this breaking change to `SatPopoverAnchor`, I was able to simplify the codebase and remove the notification service. Now the SatPopoverAnchorService is provided in the SatPopover component. the SatPopover component injects the anchor service directly.

When a SatPopover's anchor input is changed, the popover dynamically repositions itself accordingly. This provides a partial solution to #84, as you can associate multiple buttons with a single popover by updating the popovers anchor when a button is clicked.

I've updated the tests as well as the readme accordingly.

I'm open to working to get this PR into a mergeable state, if desired.

thefliik commented:
If this PR ends up getting merged, it would probably be worthwhile to more completely address #84 by adding a follow up change to `SatPopoverAnchor` to allow multiple anchors to be associated with a single popover.

thefliik commented on src/lib/popover/popover.component.ts:
```diff
@@ -223,48 +259,54 @@ export class SatPopover implements OnInit, OnDestroy {
 
   constructor(
     private _focusTrapFactory: FocusTrapFactory,
+    private _anchoringService: SatPopoverAnchoringService,
```
Note, where before the `SatPopoverAnchor` made the `_anchoringService` available publicly, here I made it private. It wasn't obvious to me why it was public in the first place, other than perhaps facilitating tests. I added a workaround to the tests, which I'll highlight below, to allow this to remain private.

thefliik commented on src/lib/popover/popover.component.ts:
```diff
@@ -223,48 +259,54 @@ export class SatPopover implements OnInit, OnDestroy {
+  setCustomAnchor(
+    viewContainer: ViewContainerRef,
+    el: ElementRef<HTMLElement> | HTMLElement
+  ): void {
```
One of the tests allowed programmatically setting a custom anchor element. Given that the SatPopover now allows setting the anchor via its `anchor` property, its not clear to me that this functionality is still needed. For the time being, I've kept the functionality by providing this method (which is the only way to programmatically set the `ViewContainerRef` associated with the popover.

thefliik commented on src/lib/popover/popover.spec.ts:
```diff
@@ -1068,13 +1107,34 @@ describe('SatPopover', () => {
+ */
+// @ts-ignore typescript doesn't like that we made prop private -> public
+export class OpenAPISatPopover extends SatPopover {
+  _anchoringService: SatPopoverAnchoringService;
```
I created this class to allow accessing a SatPopovers `_anchoringService` for tests. A popover can be typed using this class to allow accessing the `_anchoringService` without errors.

willshowell commented:
On the surface this sounds great! Thank you so much for the contribution! I agree on -- and am most interested in -- the simplification this provides to the codebase and api.

I'll look over this as I have time and get back to you 😁

pgkehle commented:
Hello @thefliik. This is the exact usage I was hoping this library could allow, as I have a need for it, specifically so one can build a component around the popover, and pass the anchor into the component. 

I also notice that the package has vulnerabilities, so I'd like to get that up to date as well.

pgkehle commented:
I've pushed updates to this library for package updates.  Then I cloned your project and rebased to `origin/master`.  After rebuilding, I see this error:

```
ERROR Error: inject() must be called from an injection context
```

For some background: 
I have not yet published a release of this library with the package updates.  As such, I had to do a soft-link hack locally to test the package updates.  

I notice these during compile time:

```
WARNING in ../thefliik-popover/node_modules/@angular/core/fesm5/core.js 18359:15-36
Critical dependency: the request of a dependency is an expression

WARNING in ../thefliik-popover/node_modules/@angular/core/fesm5/core.js 18371:15-102
Critical dependency: the request of a dependency is an expression
```

Can you please:
* Rebase your PR to `master`
* Update the `demos`
* See if you can reproduce those errors and compile warnings

Thank you!

pgkehle commented:
UPDATE 2:

I've been able to make this work as expected with our client project.   Please update the demos and rebase, and I'll merge it in and do a release. Thanks!

thefliik commented:
Regarding the merge conflict: `tslint.json` had a syntax error (invalid "," after `linterOptions.exclude`) which was preventing prettier from working on my computer.

thefliik commented:
Also note: formatting with prettier after rebasing off master has caused a lot of white noise in this PR

thefliik commented:
@pgkehle Ok, I've rebased off master and updated the demos to use the new API. Also, a few additional changes:

1. I can't remember if I knew this when I first created this PR, but there is no way to select the `HTMLElement` associated with an angular component using a template reference variable (by default, the angular component is selected or you can manually select another exported directive on the component). This contracts with a template reference variable placed on a standard html element, which automatically selects the `HTMLElement`. This means that if you want a popover anchor to be an angular component (which I expect will be common) you (generally) will need to make use of the `SatPopoverAnchor` directive. Because of this potentially confusing detail, I think it's wise to continue using satPopoverAnchor in the documentation, even for examples where it isn't actually necessary.
  - I updated this PR's description (above) with some more info on this topic which might explain it better.
2. I found I needed to add back in the `satPopoverAnchorFor` option in order to make the "anchor-reuse" demo work. So now `SatPopoverAnchor` has an `@Input()` setter called `satPopoverAnchorFor` which can be used to associate the anchor with a popover. This can be used instead of passing an anchor to a popover via the `SatPopover#anchor` input property.
3. Previously, I had made the `SatPopover#_anchoringService` private, and used a special interface in the tests to allow access to the `_anchoringService`. However, recently I learned about the `stripInternal` typescript compiler option which solves this problem more elegantly. If you are not familiar, `stripInternal: true` instructs the typescript compiler to remove the typings for methods marked with an `@internal` jsdoc comment from the output type declarations file. I enabled `stripInternal: true`, made `SatPopover#_anchoringService` public again, and added an `@internal` comment to instruct the typescript compiler to remove that property from the type declarations file (effectively making it private for the public).

thefliik commented:
One final thought: while `<div satPopoverAnchor [satPopoverAnchorFor]='p'></div>` is pretty clear (and is the way I've set up the optional `satPopoverAnchorFor` input setter currently), SatPopoverAnchor could be updated so that the satPopoverAnchorFor option became satPopoverAnchor with usage like `<div [satPopoverAnchor]='p'></div>` -- which is less verbose. Any preference?

pgkehle commented:
> One final thought: while `<div satPopoverAnchor [satPopoverAnchorFor]='p'></div>` is pretty clear (and is the way I've set up the optional `satPopoverAnchorFor` input setter currently), `SatPopoverAnchor#satPopoverAnchorFor` could be renamed to `satPopoverAnchor` with usage like `<div [satPopoverAnchor]='p'></div>` -- which is less verbose. Any preference?
> 
> Personally, I think I prefer `<div [satPopoverAnchor]='p'></div>`. But again, it's not quite as clear as the current setup.

I prefer simplicity as much as possible.

pgkehle commented:
@thefliik Just a quick statement, thank you for all of the work you've contributed to this project.  

It might be somewhat apparent, but I am picking up work on this, while multitasking in the primary job function at NC State.  My time constraints allow for maybe 5% of my time on this, so I have more to learn about the library.  I will likely lean to suggestions from those who are more familiar with the library, such as yourself.  Most of my time ends up being focused on the back end with a MongoDB graph database and corresponding ETL engine I wrote over the last couple of years, which I hope to publish to our Organization relatively soon. 

I've pushed the quick fix for the tslint.json in #157.

pgkehle commented:
RE: prettier, my apologies on the additional work.  Just when I realized that vscode was complaining so much b/c of my other projects, it was clear that the prettier requirement wasn't applied to the library.  I do like that tool for having consistent code being committed to the library.  Thank you for aligning.

pgkehle commented:
> @pgkehle Ok, I've rebased off master and updated the demos to use the new API. Also, a few additional changes:
> 
> 1. I can't remember if I knew this when I first created this PR, but there is no way to select the `HTMLElement` associated with an angular component using a template reference variable. By default, the underlying angular component is selected. You can also manually select another exported directive on the component, but you cannot manually select the underlying `HTMLElement`. This contrasts with a template reference variable placed on a standard html element, which automatically selects the underlying `HTMLElement`. This means that if you want a popover anchor to be an angular component (which I expect will be common) you (generally) will need to make use of the `SatPopoverAnchor` directive. Because of this potentially confusing detail, I think it's wise to continue using satPopoverAnchor in the documentation, even for examples where it isn't actually necessary.
>    
>    * I updated this PR's description (above) with some more info on this topic which might explain it better.

If memory servers, I remember seeing that sort of behavior when testing on the front end.  That makes sense.  


> 2. I found I needed to add back in the `satPopoverAnchorFor` option for the "anchor-reuse" demo. So now `SatPopoverAnchor` has an `@Input()` setter called `satPopoverAnchorFor` which can be used to associate the anchor with a popover. This can be used instead of passing an anchor to a popover via the `SatPopover#anchor` input property.

That is very helpful, thanks!


> 3. Previously, I had made the `SatPopover#_anchoringService` private, and used a special interface in the tests to allow access to the `_anchoringService`. However, recently I learned about the (undocumented?) `stripInternal` typescript compiler option which solves this problem more elegantly. If you are not familiar, `stripInternal: true` instructs the typescript compiler to remove the typings for methods marked with an `@internal` jsdoc comment from the output type declarations file. I enabled `stripInternal: true`, made `SatPopover#_anchoringService` public again, and added an `@internal` comment to instruct the typescript compiler to remove that property from the type declarations file (effectively making it private for the public).

That sounds like a good option. 


I'll reproduce my test scenario hopefully next week, and we can get this finalized and merged in.  Again, thanks a bunch for your work here.

thefliik commented:
> RE: prettier, my apologies on the additional work.

No problem at all.

>> Personally, I think I prefer <div [satPopoverAnchor]='p'></div>. But again, it's not quite as clear as the current setup.
>
> I prefer simplicity as much as possible.

To be clear, you're voting in favor of `<div [satPopoverAnchor]='p'></div>`, yes? Assuming yes, I'll take one final pass and update this PR to that syntax early next week. Then this should be ready to merge.

thefliik commented:
> It might be somewhat apparent, but I am picking up work on this, while multitasking in the primary job function at NC State. My time constraints allow for maybe 5% of my time on this, so I have more to learn about the library.

No worries. Currently a lot of my excess time has been spent building out a [recurring events/dates library](https://gitlab.com/john.carroll.p/rschedule).

> Most of my time ends up being focused on the back end with a MongoDB graph database and corresponding ETL engine I wrote over the last couple of years, which I hope to publish to our Organization relatively soon.

Interesting. I'll probably take a look at that when/if you open source some of it. I'm working on a volunteer project building a volunteer coordination app. I originally built out an alpha version using [Neo4j](https://neo4j.com/) as the database before deciding that managing Neo4j instances would be just too time consuming (and possibly expensive) for a volunteer project.

I'm going through the process of transitioning the project to Firebase's Firestore, though there are a few aspects of volunteer coordination that truly produce graph data and do not map to Firestore well. At the moment those bits are being handled by CosmosDB but, maybe your project has some ideas I could make use of. Of course, given that Firestore charges by the document read, there's a good chance that handling graph data inside firestore, even if possible, wouldn't be economical.

pgkehle commented:
> > It might be somewhat apparent, but I am picking up work on this, while multitasking in the primary job function at NC State. My time constraints allow for maybe 5% of my time on this, so I have more to learn about the library.
> 
> No worries. Currently a lot of my excess time has been spent building out a [recurring events/dates library](https://gitlab.com/john.carroll.p/rschedule).
> 

I took a quick peek, and that's seems to have some promise.  I've been having to deal with scheduling using old-school cron values for determining frequency.  Thus far, it seems to do well.  But I could see possibilities of needing something different. 

> > Most of my time ends up being focused on the back end with a MongoDB graph database and corresponding ETL engine I wrote over the last couple of years, which I hope to publish to our Organization relatively soon.
> 
> Interesting. I'll probably take a look at that when/if you open source some of it. I'm working on a volunteer project building a volunteer coordination app. I originally built out an alpha version using [Neo4j](https://neo4j.com/) as the database before deciding that managing Neo4j instances would be just too time consuming (and possibly expensive) for a volunteer project.
> 

This sounds like a great tool.  I've seen a number of conferences who need volunteers, and aside from having a sign up genius initial page, followed by text messaging, the coordination usually can lack in overall effectiveness.  

> I'm going through the process of transitioning the project to Firebase's Firestore, though there are a few aspects of volunteer coordination that truly produce graph data and do not map to Firestore well. At the moment those bits are being handled by CosmosDB but, maybe your project has some ideas I could make use of. Of course, given that Firestore charges by the document read, there's a good chance that handling graph data inside firestore, even if possible, wouldn't be economical.

I'd be more than happy to compare notes!

The graph db project has evolved immensely since I started it.  But I knew I needed to have the data coming in be very flexible, with uniqueness being something again flexible, determined by rules. The goal has been to be able to easily configure rules for the nodes and links, and now a level of rules to help guide you through the databased given a starting node, ending nodes, and required paths.  Someone did once suggest neo4j, but since we were already invested in the design, changing course would have been starting over.  

We've been thinking of getting a technical writer in for a bit to help document the setup process, build process, etc.  It's close to getting to the point of open sourcing, but then I type that an I think about how many times I've said that same thing.  *grin*

pgkehle commented:
> > RE: prettier, my apologies on the additional work.
> 
> No problem at all.
> 
> > > Personally, I think I prefer <div [satPopoverAnchor]='p'>. But again, it's not quite as clear as the current setup.
> > 
> > 
> > I prefer simplicity as much as possible.
> 
> To be clear, you're voting in favor of `<div [satPopoverAnchor]='p'></div>`, yes? Assuming yes, I'll take one final pass and update this PR to that syntax early next week. Then this should be ready to merge.

Yes, that's what I meant.  Thank you.

thefliik commented:
@pgkehle `<div [satPopoverAnchor]='p'></div>` is now the API. This should be ready for merging.

thefliik commented:
>  It's close to getting to the point of open sourcing, but then I type that an I think about how many times I've said that same thing. grin

I know that feeling 😃

pgkehle commented:
Please forgive the delay.  This has not been forgotten.
